packages/runtime/container-runtime/src/pendingStateManager.ts

```diff
@@ -307,217 +315,140 @@ export class PendingStateManager implements IDisposable {
+		if (message !== undefined) {
+			const messageContent = buildPendingMessageContent(message);
+
+			// Stringified content should match
+			if (pendingMessage.content !== messageContent) {
+				this.stateHandler.close(
+					DataProcessingError.create(
+						"pending local message content mismatch",
+						"unexpectedAckReceived",
+						message,
+						{
+							expectedMessageType: JSON.parse(pendingMessage.content).type,
+						},
+					),
+				);
+				return;
+			}
 		}
 
-		// Post-processing part - If we are processing a batch then this could be the last message in the batch.
-		this.maybeProcessBatchEnd(message);
-
 		return pendingMessage.localOpMetadata;
 	}
 
 	/**
-	 * This message could be the first message in batch. If so, set batch state marking the beginning of a batch.
-	 * @param message - The message that is being processed.
-	 * @param batchStartCsn - The clientSequenceNumber of the start of this message's batch (assigned during submit)
-	 * @param pendingMessage - The corresponding pendingMessage.
+	 * Do some bookkeeping for the new batch
 	 */
-	private maybeProcessBatchBegin(
-		message: ISequencedDocumentMessage,
-		batchStartCsn: number,
-		pendingMessage: IPendingMessage,
-	) {
-		if (!this.isProcessingBatch) {
-			// Expecting the start of a batch (maybe single-message).
-			if (pendingMessage.batchIdContext.batchStartCsn !== batchStartCsn) {
-				this.logger?.sendErrorEvent({
-					eventName: "BatchClientSequenceNumberMismatch",
-					details: {
-						processingBatch: !!this.pendingBatchBeginMessage,
-						pendingBatchCsn: pendingMessage.batchIdContext.batchStartCsn,
-						batchStartCsn,
-						messageBatchMetadata: (message.metadata as any)?.batch,
-						pendingMessageBatchMetadata: (pendingMessage.opMetadata as any)?.batch,
-					},
-					messageDetails: extractSafePropertiesFromMessage(message),
-				});
-			}
-		}
+	private onLocalBatchBegin(batch: InboundBatch) {
+		// Get the next message from the pending queue. Verify a message exists.
+		const pendingMessage = this.pendingMessages.peekFront();
+		assert(
+			pendingMessage !== undefined,
+			"No pending message found as we start processing this remote batch",
+		);
 
-		// This message is the first in a batch if the "batch" property on the metadata is set to true
-		if ((message.metadata as IBatchMetadata | undefined)?.batch) {
-			// We should not already be processing a batch and there should be no pending batch begin message.
+		// This could be undefined if this batch became empty on resubmit
+		const firstMessage = batch.messages.length > 0 ? batch.messages[0] : undefined;
+		if (firstMessage === undefined) {
 			assert(
-				!this.isProcessingBatch && this.pendingBatchBeginMessage === undefined,
-				0x16b /* "The pending batch state indicates we are already processing a batch" */,
+				asEmptyBatchLocalOpMetadata(pendingMessage.localOpMetadata)?.emptyBatch === true,
+				"Expected empty batch",
```

Comment:
Shouldn't this assert say `"Unexpected empty batch"`? i.e. We see the batch is empty, but the metadata says otherwise.

Reply:
No it's correct. We stick that "emptyBatch" placeholder in PendingMessages for this very moment - so when we are processing an empty inbound batch, we can correlate it with the one we sent.

Reply:
Updated the comment, does it help?

Reply:
Also, I realized we have this assert in two places.  I'm leaving it because I think it helps explain/illustrate what's going on (at least you know to ask about it, rather than just glossing over the empty case)

Reply:
What do you think about the assert message being `"Expected emptyBatch marker"`?

Reply:
Will do it in the next PR

Reply:
Seems redundant having to check for empty batch here and processPendingLocalBatch. I prefer staying DRY than having illustrative code that could be replace by more commenting. That way we could remove all empty batch handling from processPendingLocalBatch.

Reply:
I'd rather keep it there and remove it here.  We have to call `processNextPendingMessage` there anyway, and it makes a lot of sense to me to assert on the `localOpMetadata` it returns

Reply:
You could also handle this case inside the local if on processInboundBatch. That way we know the batch is not empty when it goes to processPendingLocalBatch and we don't have to worry about checking length

Reply:
Interesting thought... looked into it, but I want `onLocalBatchBegin` to run for empty and non-empty, so leaving the flow as-is.  If I misunderstood your suggestion ping me offline.

I'm just going to remove this redundant assert. Batch start validation doesn't need to worry about it since empty batch logic does.

Reply:
@kian-thompson  I also added more info to this comment.  Not the first place someone will look to learn about empty batches, but it's something.